### PR TITLE
[DTensor] StridedShard support uneven sharding

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150887
* #150862
* #150650
* __->__ #150490

This enables using FSDP+TP on parameters with dimensions that aren't
evenly divisible by the DP/TP mesh sizes.

- this may not support all possible combinations of strided shardings
  and shardings, but the support before this PR is not complete anyway

This contains several fixes for different aspects of DTensor behavior
relating to uneven strided sharding:
- original creation of the strided tensor requires fixes in
  StridedShard._split_tensor
- full_tensor() reconstruction requries fixes in
  StridedShard._to_replicate_tensor to correctly reshuffle the data into
  the original pre-sharded order
- Distributed Checkpointing support requires correct computation of the
  compute_local_shape_and_global_offset util so it knows how a local
  shard maps to the global tensor, for reconstruction during
  load/reshard.

This PR also adds a util `_explicit_order_placements` which converts a list of
placements with StridedSharding into a list of placements with only
regular sharding, with the order shuffled such that it is equivalent.

Builds on and completes the work started in https://github.com/pytorch/pytorch/pull/148894

Uneven Sharding Example
-------
(copied from _StridedShard._to_replicate_tensor docstring)

mesh = (DP=2, TP=2)
original = torch.arange(5)

**Applying Sharding**

Step 1 - Apply TP sharding
`tp = distribute_tensor(x, world_mesh['tp'], [Shard(0)])`

local_tensors:
rank0: [0,1,2]    rank1: [3,4]
rank1: [0,1,2]    rank3: [3,4]

Step 2 - Apply FSDP sharding
`dp_tp = ...` (the process of creating a strided-shard tensor is skipped over as it is hacky and complicated)
dp_tp has placement (_StridedShard(0, split_factor=2), Shard(0))
local_tensors:
rank0: [0,1]  rank1: [3]
rank1: [2]    rank3: [4]

**Reconstructing the Full Tensor**
Now, say someone wants to reconstruct dp_tp's full tensor. This will invoke 'redistribute' to replicate.
redistribute will first replicate the "Shard(0)" placement on the rightmost mesh dim, then replicate the
StridedShard placement second, which is implemented by this function.
So our starting point (`local_tensor` arg) is the result of replicating the Shard(0) placement across the
TP dim, which looks like this.

Note the discrepancy with the 'tp sharded tensor' line above!  We'll fix it by locally shuffling data.

local_tensors:
rank0: [0,1,3]  rank1: [0,1,3]
rank1: [2,4]    rank3: [2,4]

Step 1: replicate over the DP dimension.  Afterwards, each rank can locally sort the values.
  note: we need padding to do this allgather, and we'll need to keep track of the padding amount for later
	local_tensors:
rank0: [0,1,3,2,4]    rank1: [0,1,3,2,4]
rank1: [0,1,3,2,4]    rank3: [0,1,3,2,4]

Step 2: chunk and shuffle values around to account for the wrong order of operations above
and get the original tensor content back

01324#       <- our allgather includes padding, if padding was applied in step 1
01324        <- Remove the padding
013, 24      <- chunk once, 'undoing' the DP allgather
01, 3, 2, 4  <- chunk each chunk, 'undoing' the initial (wrong) TP allgather performed by Shard(0)->Replicate()
012, 34      <- interleave with stride=TP mesh dim size
01234        <- concatenate

Co-authored-by: Luca Wehrstedt <lw@meta.com>
Co-authored-by: Will Constable <whc@meta.com>

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 1]
```diff
@@ -1155,9 +1155,7 @@ def test_2d_mlp_with_nd_mesh(self):
             {
                 "reshard_after_forward": [False, True],
                 "use_activation_checkpointing": [False, True],
-                # TODO: change "mlp_dim" back to [3, 16, 17] when uneven sharding
-                # is supported for FSDP+TP
-                "mlp_dim": [4, 16, 20],
+                "mlp_dim": [5, 16, 17],
```
Comment (XilunWu) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:1158]: add `3` to the list

File: test/distributed/checkpoint/fsdp/test_fsdp_dsd.py [hunk 2]
```diff
@@ -458,39 +458,37 @@ def _test_save_with_fsdp2_tp_and_load_with_tp(
...
+        mlp_dim = 5
 
-        def _get_base_model(mlp_dim: int = 2):
-            base_model = nn.Sequential(MLP(mlp_dim), MLP(mlp_dim), MLP(mlp_dim))
+        def _get_base_model(mlp_dim: int = mlp_dim):
+            base_model = nn.Sequential(
+                MLP(mlp_dim, dim_multiplier=1),
+                MLP(mlp_dim, dim_multiplier=1),
+                MLP(mlp_dim, dim_multiplier=1),
+            )
```
Comment (XilunWu) [inline test/distributed/checkpoint/fsdp/test_fsdp_dsd.py:468]: any reason we change the model hyper-parameter instead of passing different `mlp_dim` value to test uneven sharding?

File: test/distributed/checkpoint/fsdp/test_fsdp_dsd.py [hunk 3]
```diff
@@ -521,7 +519,7 @@ def _get_base_model(mlp_dim: int = 2):
                 fsdp2_tp_optim = torch.optim.AdamW(fsdp2_tp_model.parameters(), lr=0.1)
 
                 # one-step training to modify state dict
-                inp = torch.randn((2,), device=self.rank)
+                inp = torch.randn((mlp_dim,), device=self.rank)
```
Comment (XilunWu) [inline test/distributed/checkpoint/fsdp/test_fsdp_dsd.py:522]: I see, so the test was only testing the case `mlp_dim=2`? Shall we change the line above `base_model = _get_base_model().cuda()` to `base_model = _get_base_model(mlp_dim).cuda()`, where `mlp_dim` is a variable? Right now it's fixed `mlp_dim=5`. 

It doesn't have to be in this PR, you can add a TODO

File: torch/distributed/tensor/_utils.py [hunk 4]
```diff
@@ -82,39 +82,9 @@ def compute_local_shape_and_global_offset(
...
-    a placements of [Shard(0)]. We will not have non-empty local tensor for all the ranks.
-    The local shape and global offset will be as follows:
-    rank0 -- local_shape:[1,], global_offset:[0,]
-    rank1 -- local_shape:[1,], global_offset:[1,]
-    rank2 -- local_shape:[0,], global_offset:[2,]
-    rank5 -- local_shape:[0,], global_offset:[2,]
-    rank3 -- local_shape:[0,], global_offset:[2,]
-    rank4 -- local_shape:[0,], global_offset:[2,]
-    rank6 -- local_shape:[0,], global_offset:[2,]
-    rank7 -- local_shape:[0,], global_offset:[2,]
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/tensor/_utils.py:116]: let me replace it with my example of uneven sharding 01234? i removed it bc it was IMO incorrect and pretty wordy, but also seems a bit obvious to me. i think i can make a useful and short comment.

File: torch/distributed/tensor/_utils.py [hunk 5]
```diff
@@ -82,39 +82,9 @@ def compute_local_shape_and_global_offset(
...
     """
+    ordered_placements = _explicit_order_placements(mesh.shape, placements)
```
Comment (XilunWu) [inline torch/distributed/tensor/_utils.py:86]: nice, this simplification is fantastic!

File: torch/distributed/tensor/placement_types.py [hunk 6]
```diff
@@ -134,6 +117,11 @@ def _local_shard_size_on_dim(
     ) -> tuple[int, int]:
         """
         returns the local shard size and offset on a given tensor dim
+
+        # TODO(whc)
+         - `size_on_dim` arg has a pretty confusing name, would 'global_dim_size' be better?
```
Comment (XilunWu) [inline torch/distributed/tensor/placement_types.py:122]: `global_dim_size` implies the dim size of the global tensor. However, the `size_on_dim` is the local dim size before being further sharded. Think of 2D sharding, the `size_on_dim` when sharding on FSDP is `global_dim_size // tp_size`.

File: torch/distributed/tensor/placement_types.py [hunk 7]
```diff
@@ -134,6 +117,11 @@ def _local_shard_size_on_dim(
...
+         - `local_shard_size_on_dim` could be renamed to `local_shard_size`?
```
Comment (XilunWu) [inline torch/distributed/tensor/placement_types.py:123]: agree that we can remove the `on_dim` postfix

File: torch/distributed/tensor/placement_types.py [hunk 8]
```diff
@@ -134,6 +117,11 @@ def _local_shard_size_on_dim(
...
+         - why do we have 'return_offset` bool? if we always return a tuple anyway, why not always return the real offset?
```
Comment (XilunWu) [inline torch/distributed/tensor/placement_types.py:124]: yes we should remove `return_offset`

**Review by XilunWu — COMMENTED** [review]
overall LGTM except the `_to_replicate_tensor` function. Need more time to understand it.

File: torch/distributed/tensor/_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/tensor/_utils.py:116]: I see. I don't think we need to `01234` example here because basically we want to express what `compute_local_shape_offset` does via the comment. IMO a regular uneven sharding (i.e. no stride) would be sufficient.

**Comment by XilunWu** [comment]
oops just realized https://github.com/pytorch/pytorch/pull/150354 wasn't merged. Now merging it, please rebase ;-)

File: test/distributed/checkpoint/fsdp/test_fsdp_dsd.py (hunk 2 above)
Comment (wconstab) [reply test/distributed/checkpoint/fsdp/test_fsdp_dsd.py]: do you mean why do i change the 'dim_multiplier' instead of just changing the mlp_dim? or ?

File: torch/distributed/tensor/placement_types.py (hunk 6 above)
Comment (wconstab) [reply torch/distributed/tensor/placement_types.py:122]: moved these changes to a new PR on the stack and removed the TODOs from here

File: torch/distributed/tensor/placement_types.py [hunk 9]
```diff
@@ -508,17 +488,77 @@ def _to_replicate_tensor(
...
+        # single-gpu "weight" of size 5, will be 'uneven' for sharding
+        original = torch.arange(5)
+
+        tp sharded tensor
+        -----------------
+        `tp = distribute_tensor(x, world_mesh['tp'], [Shard(0)])`
+
+        local_tensors:
+        rank0: [0,1,2]    rank1: [3,4]
+        rank1: [0,1,2]    rank3: [3,4]
```
Comment (wanchaol) [inline torch/distributed/tensor/placement_types.py:513]: rank2?

File: torch/distributed/tensor/placement_types.py [hunk 10]
```diff
@@ -508,17 +488,77 @@ def _to_replicate_tensor(
...
+
+        fsdp+tp sharded tensor
+        ----------------------
+        `dp_tp = ...` (the process of creating a strided-shard tensor is skipped over as it is hacky and complicated
+        #TODO put an example somewhre and ref to it)
+        dp_tp has placement (_StridedShard(0, split_factor=2), Shard(0))
+        local_tensors:
+        rank0: [0,1]  rank1: [3]
+        rank1: [2]    rank3: [4]
```
Comment (wanchaol) [inline torch/distributed/tensor/placement_types.py:522]: rank2

**Review by wanchaol — COMMENTED** [review]
The approach looks good to me in general, have some comments inlined

File: torch/distributed/tensor/placement_types.py [hunk 11]
```diff
@@ -186,14 +169,17 @@ def _shard_tensor(
...
         # perform scatter from the src_data_rank as data source when it is not None
         mesh_scatter(
             output, scatter_list, mesh, mesh_dim=mesh_dim, group_src=src_data_rank
         )
 
         # Only unpad if the local_tensor was padded on the dimension.
-        if pad_sizes and pad_sizes[mesh_dim_local_rank] > 0:
+        if pad_sizes[mesh_dim_local_rank] > 0:
             output = unpad_tensor(output, self.dim, pad_sizes[mesh_dim_local_rank])
+            # Unpad might return a view, hence we need to remake it contiguous
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/tensor/placement_types.py:181]: This part came from @lw's initial PR, here is an explanation:
https://github.com/pytorch/pytorch/pull/148894#discussion_r1989121982

File: test/distributed/tensor/test_utils.py [hunk 12]
```diff
@@ -191,6 +191,31 @@ def test_fsdp_tp_meta_compute(self):
...
+        global_mesh = init_device_mesh(
+            self.device_type, (dp_size, tp_size), mesh_dim_names=("dp", "tp")
+        )
+        global_tensor_shape = torch.Size([15, 5])
+        placements = [_StridedShard(0, split_factor=tp_size), Shard(0)]
+        local_shape, global_offset = compute_local_shape_and_global_offset(
+            global_tensor_shape, global_mesh, placements
+        )
+        rank = global_mesh.get_rank()
+        if dp_size == 2:
```
[the remark this answers is not in the document]
Comment (wconstab) [reply test/distributed/tensor/test_utils.py:215]: yea, good point. I think i can just remove this case.  I may have been testing locally with only 4 gpus at one point, but as long as CI has 8 and we are sure it always runs with 8, that is fine.

**Review by wanchaol — APPROVED** [review]

File: test/distributed/checkpoint/fsdp/test_fsdp_dsd.py (hunk 2 above)
Comment (XilunWu) [reply test/distributed/checkpoint/fsdp/test_fsdp_dsd.py]: right. the default value of `dim_multiplier` is 4. Why are we changing the value here? Is it to create an uneven sharding case? Then we better keep the `mlp_dim=2, dim_multiplier=4` case by parameterizing `dim_multiplier` in test in `[1, 4]`. WDYT? We can add it in another PR to avoid wasting CI.

**Review by XilunWu — APPROVED** [review]
LGTM! Thanks for addressing all the comments. Stamp to unblock. One suggestion is we improve `test_fsdp_dsd` test coverage in a followup PR because this PR removes the even sharding case.
